### PR TITLE
fix(scanner): use mTLS between Matcher and Indexer

### PR DESCRIPTION
**Pull request by RTann**
## Description

Now that ROX-19050 is complete, it's time to use mTLS between Indexer and Matcher. This is done via `clientconn.AuthenticatedGRPCConnection`.

To get this to work for Scanner v4's use-case, this PR adds `ServerName` as an option for `clientconn.AuthenticatedGRPCConnection` and now requires users to specify a context instead of using `context.Background()` for everything. At this time, all non-Scanner v4 uses of this function simply use `context.Background()` to ensure no changes.

## Checklist
- [x] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

### Here I tell how I validated my change

1. Deploy Central Services
2. Set `ROX_API_TOKEN`, if needed
3. `roxctl --insecure-skip-tls-verify image scan -i quay.io/stackrox-io/main:4.3.4`
4. Verify it succeeds

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by fredrb — APPROVED** [review]
Checked the sensor/gRPC part and looks good to me.

File: pkg/scannerv4/client/client.go
Comment (dcaravel) [inline pkg/scannerv4/client/client.go:109]: Glad you removed this, was planning on creating a follow up item, as it didn't seem the client needed access to the root CA / key.

**Review by dcaravel — COMMENTED** [review]
Changes LGTM, holding off on approve until we test it w/ release builds.

Wondering too if we should change the scanner `AuthFuncOverride` to not rely on 'release' builds to test auth.  Thinking we should by default, on any build, have auth enabled - and instead add an environment var, config, etc. to disable auth when needed so that we default to secure  - this would be similar to what we had with scanner v2's `ROX_SKIP_PEER_VALIDATION`

File: pkg/scannerv4/client/client.go
Comment (RTann) [reply pkg/scannerv4/client/client.go:109]: client does need access to the root CA since we need to trust a non-builtin trusted CA, Central. My first iteration of this still required the CA's key, which was wrong, though. Now we don't require the key :)

**Comment by RTann** [comment]
> Changes LGTM, holding off on approve until we test it w/ release builds.

I tested with the release build from this PR (added the `ci-release-build` label to build a `release` build). I do want to test again before merging, though, just to confirm again

> Wondering too if we should change the scanner `AuthFuncOverride` to not rely on 'release' builds to test auth. Thinking we should by default, on any build, have auth enabled - and instead add an environment var, config, etc. to disable auth when needed so that we default to secure - this would be similar to what we had with scanner v2's `ROX_SKIP_PEER_VALIDATION`

I think we can make this more configurable in the future. Also, `ROX_SKIP_PEER_VALIDATION` worked fine, though one problem I had with it was that it allowed us to skip this validation even in release builds, which we didn't want. We don't really want anyone using this in production to be able to turn off client-side validation

File: pkg/scannerv4/client/client.go
Comment (RTann) [reply pkg/scannerv4/client/client.go:109]: the key thing was an issue when `mtls.LoadDefaultCA()` was used in `createGRPCConn`

**Comment by dcaravel** [comment]
> I tested with the release build from this PR (added the ci-release-build label to build a release build). I do want to test again before merging, though, just to confirm again

I didn't see the `-prerelease` tag created in quay (or the GH status showing prerelease was built), something I'm missing?

image

**Comment by RTann** [comment]
> > I tested with the release build from this PR (added the ci-release-build label to build a release build). I do want to test again before merging, though, just to confirm again
> 
> I didn't see the `-prerelease` tag created in quay (or the GH status showing prerelease was built), something I'm missing?
> 
> image

yeah that's prerelease, which we'd enabled with `ci-build-prerelease`. I used `ci-release-build` which will output the normal tag but with the go tags set to `release`. See https://github.com/stackrox/stackrox/actions/runs/7672793453/job/20914088099?pr=9555

What's the purpose of `prerelease`? I'm not sure. I just know it exists

**Review by dcaravel — APPROVED** [review]

**Pull request merged**
